### ISSUE TITLE
html_dart2js.dart readEntries() fails in Dart 2.0.0-dev.69.0

### ISSUE THREAD

**Issue by tonyclickspace** [dart-lang/sdk#33891]
Dart VM version: 2.0.0-dev.69.0 (Tue Jul 17 14:57:16 2018 +0200) on "macos_x64", running in Chrome Version 68.0.3440.42 (Official Build) beta (64-bit).

Compiled to javascript using dart2js using `--csp` (not sure if that matters)

```
void main() {
  window.navigator.persistentStorage.requestQuota(1024*1024, _quotaHandler);
}
Future _quotaHandler(int byteCount) async {
  FileSystem filesystem = await window.requestFileSystem(1024*1024, persistent:true);
  DirectoryEntry dir = await filesystem.root.getDirectory('/');
  DirectoryReader dirReader = dir.createReader();
  await dirReader.readEntries(); //error
}
```

Throws
`Uncaught Error: TypeError: Instance of 'JSArray': type 'JSArray' is not a subtype of type 'List<Entry>'`

Looks like `readEntries` is missing a conversion step that wasn't necessary in Dart 1.

**Comment by sigmundch** [comment]
Thanks for the bug report and for the beautifully short repro example!

It seems that the html library has an invalid type internally, and that triggers this assertion failure.

html_dart2js.dart is a too big to display in github, so here are the relevant pieces:
```dart
...
typedef void _EntriesCallback(List<Entry> entries); // [a]
...
@Native("DirectoryReader")
class DirectoryReader extends Interceptor {
  // To suppress missing implicit constructor warnings.
  factory DirectoryReader._() {
    throw new UnsupportedError("Not supported");
  }

  @JSName('readEntries')
  void _readEntries(_EntriesCallback successCallback,
      [_ErrorCallback errorCallback]) native;

  Future<List<Entry>> readEntries() {
    var completer = new Completer<List<Entry>>();
    _readEntries((value) { // [b]
      completer.complete(new List<Entry>.from(value));  // [c]
```

See how [a] declares the callback as a closure expecting a `List<Entry>`, doing so makes the compiler infer that the closure in [b] expects a `List<Entry>` argument.

At this time dart2js doesn't optimize away Dart2 checks (even though there are flags like `--omit-implicit-checks` to skip them), so it generates a check at the entry of every closure to ensure that their inputs are valid.

However, the browser here doesn't return a `List<Entry>`, but a `List<dynamic>` (JSArray). Which is why there is a copy in [c] to convert the list to the proper type.

The correct fix would be to change [a] from `typedef void _EntriesCallback(List<Entry> entries);` to `typedef void _EntriesCallback(List entries);`.

@terrylucas - I was able to verify that this fixes the issue if I update html_dart2js.dart directly. However, I am not familiar with how to change the code generator to properly fix this. Can you take a look?

**Comment by tonyclickspace** [comment]
If the solution is fixing the code generator, there may be similar undiscovered issues. Opening up html_dart2js.dart and looking for `typedef void`, I found a few items that mention lists in their arguments:

```
typedef void IntersectionObserverCallback(
    List<IntersectionObserverEntry> entries, IntersectionObserver observer);
typedef void MutationCallback(
    List<MutationRecord> mutations, MutationObserver observer);
typedef void PerformanceObserverCallback(
    PerformanceObserverEntryList entries, PerformanceObserver observer); // maybe?
typedef void ReportingObserverCallback(
    List<_Report> reports, ReportingObserver observer);
typedef void ResizeObserverCallback(
    List<ResizeObserverEntry> entries, ResizeObserver observer);
```

This might need to be addressed before Dart2Stable...? cc @kevmoo

**Comment by morrica** [comment]
It looks like this issue was not fully resolved by 7b77fa2. As @tonyclickspace says (https://github.com/dart-lang/sdk/issues/33891#issuecomment-406376399), `ResizeObserverCallback` looks like it needs the fix, however a change for it is not included in the commit.

I believe this is causing the issue referenced here: https://github.com/dart-lang/sdk/issues/36798#issuecomment-557162660

Can this issue be reopened or should I file a new one?

**Comment by sigmundch** [comment]
You are correct, thanks for following up. We can continue tracking this in #36798
